OADP-398: API Reference (#47791)
openshift/openshift-docs: modules/oadp-api-details.adoc

// Module included in the following assemblies:
//
// * oadp-api

:_content-type: REFERENCE
[id="oadp-api-details_{context}"]
= OADP API

The following tables expose the structure of the OADP API:

.DataProtectionApplicationSpec
[options="header"]
|===
|Property|Type|Description

|`BackupLocations`

> |[] `BackupLocation`

Reviewer: maybe add a link [][BackupLocation](https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#BackupLocation) or  hyperlink to Table 2. BackupLocation

Author: Done.

|Defines the list of desired configurations to use for `BackupStorageLocations`.

|`SnapshotLocations`

> |[] `SnapshotLocation`

Reviewer: maybe add a link [][SnapshotLocation](https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#SnapshotLocation) or hyperlink to `Table 3. SnapshotLocation`

Author: Done.

|Defines the list of desired configurations to use for `VolumeSnapshotLocations`.

|`UnsupportedOverrides`

> |map [ `UnsupportedImageKey` ] string

Reviewer: maybe add as a link [UnsupportedImageKey](https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#UnsupportedImageKey)

Author: Done.

|Can be used to override the deployed dependent images for development.

|`PodAnnotations`
|map [ string ] string
|Used to add annotations to pods deployed by Operators.

|`PodDnsPolicy`
|[`corev1.DNSPolicy`] link:https://pkg.go.dev/k8s.io/api/core/v1#DNSPolicy[DNSPolicy]
|Defines how the configuration of the DNS of a pod.

|`PodDnsConfig`
|[`corev1.PodDNSConfig`] link:https://pkg.go.dev/k8s.io/api/core/v1#PodDNSConfig[PodDNSConfig]
|Defines the DNS parameters of a pod in addition to those generated from `DNSPolicy`.

|`BackupImages`
|*bool
|Used to specify whether or not you want to deploy a registry for enabling backup and restore of images.

|`Configuration`
|*`ApplicationConfig`
|Used to define the data protection application's server configuration.

> |===

Reviewer 2: Probably missed data mover because our reference docs in https://github.com/openshift/oadp-operator/blob/master/docs/API_ref.md still isn't updated. @shubham-pampattiwar @eemcmullan @savitharaghunathan please update upstream as well. Once this PR is reviewed we could consider redirecting everything to downstream.

> |===

Author: Added.

.BackupLocation
[options="header"]
|===
|Property|Type|Description

|`name`
|`metav1. ObjectMeta`
|Location name.

> |`velero`

Reviewer: **V**elero

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L101 `velero` is correct.

> |*link:(https://velero.io/docs/v1.9/api-types/backupstoragelocation/[velero.BackupStorageLocationSpec]

Reviewer: type should be *[velero](https://pkg.go.dev/github.com/vmware-tanzu/velero/pkg/apis/velero/v1).[BackupStorageLocationSpec](https://pkg.go.dev/github.com/vmware-tanzu/velero/pkg/apis/velero/v1#BackupStorageLocationSpec)

Author: Done.

|Location to store volume snapshots. For further details, see link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/bsl_and_vsl.md[Backup Storage Locations and Volume Snapshot Locations Customization].
|===

> .VolumeSnapshot

Reviewer: Should be SnapshotLocation instead of VolumeSnapshot

Author: @mperetzred -- Is the VolumeSnapshot table in https://github.com/openshift/oadp-operator/blob/master/docs/API_ref.md  incorrect?

Reviewer 2: Correct @RichardHoch 
Per https://pkg.go.dev/github.com/openshift/oadp-operator@v1.0.3/api/v1alpha1#DataProtectionApplicationSpec it is now snapshot location. Needs updating in upstream.

Reviewer 2: Created issue https://github.com/openshift/oadp-operator/issues/768

[options="header"]
|===
|Property|Type|Description

|`name`
|`metav1. ObjectMeta`

> |Location name.

Author: @mperetzred -- see queston above.

> |`velero`

Reviewer: **V**elero (CamelCase, not snakeCase)

Author: Done.

Reviewer 2: Similar to https://github.com/openshift/openshift-docs/pull/47791#discussion_r928266288

`velero` is correct.

> |*link:(https://velero.io/docs/v1.9/api-types/volumesnapshotlocation/[velero.VolumeSnapshotLocationSpec]

Reviewer: maybe point to https://pkg.go.dev/github.com/vmware-tanzu/velero/pkg/apis/velero/v1#VolumeSnapshotLocationSpec instead

|Location to store volume snapshots. For further details, see link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/bsl_and_vsl.md[Backup Storage Locations and Volume Snapshot Locations Customization].
|===

> .ApplicationConfig (DataProtectionApplicationSpec.Configuration)

Reviewer: maybe add a link https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#ApplicationConfig (make .ApplicationConfig as a link)

Author: Done -- but we cannot use links in headers, so I put the link after the table.

[options="header"]
|===
|Property|Type|Description

> |`velero`

Reviewer: **V**elero (CamelCase, not snakeCase)

Reviewer 2: Per  https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L297
`velero` is correct

> |*VeleroConfig

Reviewer: maybe point to https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#VeleroConfig or a hyperlink to `Table 5. VeleroConfig`

Author: Done

|Defines the configuration for the Velero server.

> |`restic`

Reviewer: **R**estic (CamelCase, not snakeCase)

Author: Ok

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L185
restic snakeCase is correct.

> |*resticConfig

Reviewer: **R**esticConfig

Author: Done.

|Defines the configuration for the Restic server.
|===

.VeleroConfig
[options="header"]
|===
|Property|Type|Description

> |`featureFlags`

Reviewer: - **F**eatureFlags (CamelCase, not snakeCase)

- the type it's not a pointer from what I checked: it's should be just []string

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L317
`featureFlags` is correct

> |*[] string

Reviewer: the type it's not a pointer from what I checked: it's should be just []string

|Defines the list of features to enable for the Velero instance.

> |`defaultPlugins`

Reviewer: **D**efaultPlugins (CamelCase, not snakeCase)

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L313
`defaultPlugins` is correct.

> |*[] string

Reviewer: it's not of type *[]string, but [][DefaultPlugin](https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#DefaultPlugin):

Reviewer 2: To the user it's still string per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L315

Only in code is the type casted to DefaultPlugin.. but this is minor detail.

|Five types of default Velero plug-ins can be installed: `AWS`,`Azure`, `CSI`, `GCP`, and `OpenShift`.

> |`customPlugins`

Reviewer: **C**ustomPlugins

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L299
`customPlugins` is correct.

Reviewer 2: Also for fyi.. you can see the case that will be used in yaml  looking at the json tag at the end.
```
CustomPlugins []CustomPlugin `json:"customPlugins,omitempty"`
```
From https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#VeleroConfig

> |map [string]interface{}

Reviewer: it's not of type map [string]interface{}, but [CustomPlugin](https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#CustomPlugin)

Author: Done.

|Used for installation of custom Velero plug-ins.

> See link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/plugins.md[plug-ins customization] for further information on `defaultPlugins` and `customPlugins`.

Reviewer: maybe point this link the the downstream doc of the plugins instead

> See link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/plugins.md[plug-ins customization] for further information on `defaultPlugins` and `customPlugins`.

Reviewer: maybe point this link the the downstream doc of the plugins instead

Author: @mperetzred I can't do that until the relevant PR is merged.

> |`restoreResourcesVersionPriority`

Reviewer: **R**estoreResourcesVersionPriority

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L433
`restoreResourcesVersionPriority` is correct

|string
|Represents a configmap that is created if defined for use in conjunction with the `EnableAPIGroupVersions` feature flag. Defining this field automatically adds `EnableAPIGroupVersions` to the Velero server feature flag.

> |`noDefaultBackupLocation`

Reviewer: NoDefaultBackupLocation

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L335 `noDefaultBackupLocation` is correct

|bool
|if you need to install Velero without a default backup storage location, the `noDefaultBackupLocation` flag must be set in order confirm installation.

> |`podConfig`

Reviewer: **P**odConfig

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L340
`podConfig` is correct.

> |*`PodConfig`

Reviewer: maybe add as a link: *[PodConfig](https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#PodConfig) or add hypelink to `Table 7. PodConfig`

|Defies the configuration of the Velero pod.

> |===

Reviewer: missing additional property: LogLevel [string](https://pkg.go.dev/builtin#string)

.ResticConfig
[options="header"]
|===
|Property|Type|Description

> |`enable`

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L188
`enable` is correct.

|*bool
|If set to `true`, enables backup/restore using Restic. If set to `false`, snapshots are needed.

> |`supplementalGroups`

Reviewer: **S**upplementalGroups

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L285 `supplementalGroups` is correct

|[]int64
|Defines the Linux groups to be applied to the Restic pod.

> |`timeout`

Reviewer: **T**imeout

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L292
`timeout` is correct.

|string
|Defines the Restic timeout. Default value is 1 hour.

> |`podConfig`

Reviewer: **P**odConfig

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L192 `podConfig` is correct.

> |*`podConfig`

Reviewer: maybe add a link *[PodConfig](https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#PodConfig) or hyperlink to `Table 7. PodConfig`

|Defies the configuration of the Restic pod.
|===

.PodConfig
[options="header"]
|===
|Property|Type|Description

> |`nodeSelector`

Reviewer: **N**odeSelector

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L200 `nodeSelector` is correct.

|map [ string ] string
|Defines the `nodeSelector` to be supplied to a Velero or Restic `podSpec`.

> |`tolerations`

Reviewer: **T**olerations

Reviewer 2: Per https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L239 `tolerations` is correct.

> |[]int64

Reviewer: Tolatrations is of type  [][corev1](https://pkg.go.dev/k8s.io/api/core/v1).[Toleration](https://pkg.go.dev/k8s.io/api/core/v1#Toleration)

Reviewer 2: +1

|Defines the Linux groups to be applied to the Restic pod.

> |`timeout`

Reviewer: there is not timeout property

Reviewer 2: There is timeout https://pkg.go.dev/github.com/openshift/oadp-operator@master/api/v1alpha1#DataMover

|link:https://pkg.go.dev/k8s.io/api/core/v1#Toleration[corev1. Toleration]
|Defines the list of tolerations to be applied to a Velero deployment or a Restic `daemonset`.

> |===

Reviewer: missing additional properties:
ResourceAllocations | [corev1](https://pkg.go.dev/k8s.io/api/core/v1).[ResourceRequirements](https://pkg.go.dev/k8s.io/api/core/v1#ResourceRequirements) | ResourceAllocations defines the CPU and Memory resource allocations for the restic Pod

Labels | map[[string](https://pkg.go.dev/builtin#string)][string](https://pkg.go.dev/builtin#string) | Labels to add to pods

|`SnapshotLocations`
|[] `SnapshotLocation`
|Defines the list of desired configurations to use for `VolumeSnapshotLocations`.

|`UnsupportedOverrides`
|map [ `UnsupportedImageKey` ] string
|Can be used to override the deployed dependent images for development.

|`PodAnnotations`
|map [ string ] string
|Used to add annotations to pods deployed by Operators.

|`PodDnsPolicy`
|[`corev1.DNSPolicy`] link:https://pkg.go.dev/k8s.io/api/core/v1#DNSPolicy[DNSPolicy]
|Defines how the configuration of the DNS of a pod.

|`PodDnsConfig`
|[`corev1.PodDNSConfig`] link:https://pkg.go.dev/k8s.io/api/core/v1#PodDNSConfig[PodDNSConfig]
|Defines the DNS parameters of a pod in addition to those generated from `DNSPolicy`.

|`BackupImages`
|*bool
|Used to specify whether or not you want to deploy a registry for enabling backup and restore of images.

|`Configuration`
|*`ApplicationConfig`
|Used to define the data protection application's server configuration.
|===

.BackupLocation
[options="header"]
|===
|Property|Type|Description

|`name`
|`metav1. ObjectMeta`
|Location name.

|`Velero`
|*link:(https://velero.io/docs/v1.9/api-types/backupstoragelocation/[velero.BackupStorageLocationSpec]

> |Location to store volume snapshots. For further details, see link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/bsl_and_vsl.md[Backup Storage Locations and Volume Snapshot Locations Customization].

Reviewer 3: @bergerhoffer 
Does this GitHub link need to be approved?

|`podAnnotations`
|map [ link:https://pkg.go.dev/builtin#string[string] ] link:https://pkg.go.dev/builtin#string[string]
|Used to add annotations to pods deployed by Operators.

|`podDnsPolicy`
|link:https://pkg.go.dev/k8s.io/api/core/v1#DNSPolicy[`corev1.DNSPolicy`]
|Defines the configuration of the DNS of a pod.

|`podDnsConfig`
|link:https://pkg.go.dev/k8s.io/api/core/v1#PodDNSConfig[`corev1.PodDNSConfig`]
|Defines the DNS parameters of a pod in addition to those generated from `DNSPolicy`.

|`backupImages`
|*link:https://pkg.go.dev/builtin#bool[bool]
|Used to specify whether or not you want to deploy a registry for enabling backup and restore of images.

|`configuration`
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220722142130-a02487302a02/api/v1alpha1#ApplicationConfig[`ApplicationConfig`]
|Used to define the data protection application's server configuration.

|`features`

> |*link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220722142130-a02487302a02/api/v1alpha1#Features[`Features`]

Reviewer 2: Blocked until https://github.com/openshift/oadp-operator/pull/747 is merged and ideally oadp-operator repo is v1.1.0 tagged. I would want to update this link from @v0.0.0-* to @v1.1.0

Reviewer 2: Please replace all references of https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220722142130-a02487302a02 with https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd

which contains up to date [features property](https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#Features).

|Location to store volume snapshots, as described in link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/bsl_and_vsl.md[Backup Storage Locations and Volume Snapshot Locations Customization].
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#BackupLocation[Complete schema definitions for the type `BackupLocation`].

.SnapshotLocation
[options="header"]
|===
|Property|Type|Description

|`name`
|link:https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ObjectMeta[metav1.ObjectMeta]
|

|`velero`
|*link:https://velero.io/docs/v1.9/api-types/volumesnapshotlocation/[velero.VolumeSnapshotLocationSpec]
|Location to store volume snapshots, as described in link:https://github.com/openshift/oadp-operator/blob/master/docs/config/bsl_and_vsl.md[Volume Snapshot Location Specification].
|===

linkhttps://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#SnapshotLocation[Complete schema definitions for the type `SnapshotLocation`].

.ApplicationConfig
[options="header"]
|===
|Property|Type|Description

|`velero`

> |*link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#VeleroConfig[VeleroConfig]

Reviewer 2: ```suggestion
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#VeleroConfig[VeleroConfig]
```
Just for consistency, I would use the url with a specific commit.

Replace all `https://pkg.go.dev/github.com/openshift/oadp-operator/api` with
`https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api`

|===
|Property|Type|Description

|`nodeSelector`
|map [ link:https://pkg.go.dev/builtin#string[string] ] link:https://pkg.go.dev/builtin#string[string]
|Defines the `nodeSelector` to be supplied to a Velero or Restic `podSpec`.

|`tolerations`
|[]link:https://pkg.go.dev/k8s.io/api/core/v1#Toleration[ Corev1.Toleration]
|Defines the list of tolerations to be applied to a Velero deployment or a Restic `daemonset`.

|`resourceAllocations`
|link:https://pkg.go.dev/k8s.io/api/core/v1#ResourceRequirements[corev1.ResourceRequirements]
|Set specific resource `limits` and `requests` for the Velero or Restic pods as described in link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/resource_req_limits.md[Resource Requests and Limits Customization].

|`Labels`
|map [ link:https://pkg.go.dev/builtin#string[string] ] link:https://pkg.go.dev/builtin#string[string]
|Labels to add to pods.
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#PodConfig[Complete schema definitions for the type `PodConfig`].

.Features
[options="header"]
|===
|Property|Type|Description

|`DataMover`

> |`DataMover`

Reviewer 2: ```suggestion
|link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#DataMover[`DataMover`]
```

|link:https://pkg.go.dev/k8s.io/api/core/v1#DNSPolicy[`corev1.DNSPolicy`]
|Defines the configuration of the DNS of a pod.

|`podDnsConfig`
|link:https://pkg.go.dev/k8s.io/api/core/v1#PodDNSConfig[`corev1.PodDNSConfig`]
|Defines the DNS parameters of a pod in addition to those generated from `DNSPolicy`.

|`backupImages`
|*link:https://pkg.go.dev/builtin#bool[bool]
|Used to specify whether or not you want to deploy a registry for enabling backup and restore of images.

|`configuration`
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#ApplicationConfig[`ApplicationConfig`]
|Used to define the data protection application's server configuration.

|`features`
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#Features[`Features`]
|Defines the configuration for the DPA to enable the Technology Preview features.
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#DataProtectionApplicationSpec[Complete schema definitions for the OADP API].

.BackupLocation
[options="header"]
|===
|Property|Type|Description

> |`name`

Reviewer: It's not possible to specify a name for the BackupLoation at the moment...
please drop this row

|link:https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ObjectMeta[`metav1.ObjectMeta`]
|Location name.

|`velero`
|*link:https://velero.io/docs/v1.9/api-types/backupstoragelocation/[velero.BackupStorageLocationSpec]
|Location to store volume snapshots, as described in link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/bsl_and_vsl.md[Backup Storage Locations and Volume Snapshot Locations Customization].
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#BackupLocation[Complete schema definitions for the type `BackupLocation`].

.SnapshotLocation
[options="header"]
|===
|Property|Type|Description

> |`name`

Reviewer: it's not possible to specify a name at the moment, see the struct definition:
https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#SnapshotLocation
please drop this row

|Five types of default Velero plug-ins can be installed: `AWS`,`Azure`, `CSI`, `GCP`, and `OpenShift`.

|`customPlugins`
|[]link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#CustomPlugin[CustomPlugin]
|Used for installation of custom Velero plug-ins.

link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/plugins.md[Plug-ins customization] gives more details on working with default and custom plug-ins.

|`restoreResourcesVersionPriority`
|link:https://pkg.go.dev/builtin#string[string]
|Represents a configmap that is created if defined for use in conjunction with the `EnableAPIGroupVersions` feature flag. Defining this field automatically adds `EnableAPIGroupVersions` to the Velero server feature flag.

|`noDefaultBackupLocation`
|link:https://pkg.go.dev/builtin#bool[bool]
|To install Velero without a default backup storage location, you must set the `noDefaultBackupLocation` flag in order to confirm installation.

|`podConfig`
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#PodConfig[`PodConfig`]

> |Defies the configuration of the `Velero` pod.

Reviewer: *Defines, typo

|`logLevel`
|link:https://pkg.go.dev/builtin#string[string]
|Velero server’s log level (use Debug for the most granular logging, leave unset for Velero default). Valid options are `trace`, `debug`, `info`, `warning`, `error`, `fatal`, and `panic`.
|===

> link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#VeleroConfig[Complete schema definitions for the type `VeleroConfig`].

Reviewer: link is broken. Please use a link without a specific version, like:
https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#VeleroConfig maybe

.CustomPlugin
[options="header"]
|===
|Property|Type|Description

|`name`
|link:https://pkg.go.dev/builtin#string[string]
|Name of custom plug-in.

|`image`
|link:https://pkg.go.dev/builtin#string[string]
|Image of custom plug-in.
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#CustomPlugin[Complete schema definitions for the type `CustomPlugin`].

.ResticConfig
[options="header"]
|===
|Property|Type|Description

|`enable`
|*link:https://pkg.go.dev/builtin#bool[bool]
|If set to `true`, enables backup and restore using Restic. If set to `false`, snapshots are needed.

|`supplementalGroups`
|[]link:https://pkg.go.dev/builtin#int64[int64]
|Defines the Linux groups to be applied to the `Restic` pod.

|`timeout`
|link:https://pkg.go.dev/builtin#string[string]
|Defines the Restic timeout. Default value is 1 hour.

|`PodConfig`
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator@v0.0.0-20220727202820-c5f87df8fafd/api/v1alpha1#PodConfig[`PodConfig`]

> |Defies the configuration of the `Restic` pod.

Reviewer: *Defines, typo

// Module included in the following assemblies:
//
// * oadp-api.adoc

:_content-type: REFERENCE
[id="oadp-api-details_{context}"]
= OADP API

The following tables expose the structure of the OADP API:

.DataProtectionApplicationSpec
[options="header"]
|===
|Property|Type|Description

|`backupLocations`
|[] link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#BackupLocation[`BackupLocation`]
|Defines the list of configurations to use for `BackupStorageLocations`.

|`snapshotLocations`
|[] link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#SnapshotLocation[`SnapshotLocation`]
|Defines the list of configurations to use for `VolumeSnapshotLocations`.

|`unsupportedOverrides`

> |map [ link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#UnsupportedImageKey)[UnsupportedImageKey] ]  link:https://pkg.go.dev/builtin#string[string]

Reviewer: please remove the `)` at the end of the URL of https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#UnsupportedImageKey as it points to the wrong location this way

|Used to add annotations to pods deployed by Operators.

|`podDnsPolicy`
|link:https://pkg.go.dev/k8s.io/api/core/v1#DNSPolicy[`corev1.DNSPolicy`]
|Defines the configuration of the DNS of a pod.

|`podDnsConfig`
|link:https://pkg.go.dev/k8s.io/api/core/v1#PodDNSConfig[`corev1.PodDNSConfig`]
|Defines the DNS parameters of a pod in addition to those generated from `DNSPolicy`.

|`backupImages`
|*link:https://pkg.go.dev/builtin#bool[bool]
|Used to specify whether or not you want to deploy a registry for enabling backup and restore of images.

|`configuration`
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#ApplicationConfig[`ApplicationConfig`]
|Used to define the data protection application's server configuration.

|`features`
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#Features[`Features`]
|Defines the configuration for the DPA to enable the Technology Preview features.
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#DataProtectionApplicationSpec[Complete schema definitions for the OADP API].

.BackupLocation
[options="header"]
|===
|Property|Type|Description

|`velero`

> |*link:https://velero.io/docs/v1.9/api-types/backupstoragelocation/[velero.BackupStorageLocationSpec]

Reviewer: could you please change that to point to https://pkg.go.dev/github.com/vmware-tanzu/velero/pkg/apis/velero/v1#BackupStorageLocationSpec so it would be consistent with the rest of the schema link references?

> |Location to store volume snapshots, as described in link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/bsl_and_vsl.md[Backup Storage Locations and Volume Snapshot Locations Customization].

Reviewer: can we point this link to some downstream reference instead? I'm unsure if there is such doc downstream though

|`bucket`

> | [* link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#CloudStorageLocation[CloudStorageLocation] ]

Reviewer: why is `*CloudStorageLocation` between square brackets? should be just `*CloudStorageLocation` and not `[*CloudStorageLocation]`

| [Technology Preview] Automates creation of a bucket at some cloud storage providers for use as a backup storage location.

[IMPORTANT] +
==== +
The `bucket` parameter is a Technology Preview feature only. Technology Preview features are not supported with Red Hat production service level agreements (SLAs) and might not be functionally complete. Red Hat does not recommend using them
in production. These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.

For more information about the support scope of Red Hat Technology Preview features, see link:https://access.redhat.com/support/offerings/techpreview/[Technology Preview].
====
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#BackupLocation[Complete schema definitions for the type `BackupLocation`].

.SnapshotLocation
[options="header"]
|===
|Property|Type|Description

|`velero`

> |*link:https://velero.io/docs/v1.9/api-types/volumesnapshotlocation/[velero.VolumeSnapshotLocationSpec]

Reviewer: could you please change that to point to https://pkg.go.dev/github.com/vmware-tanzu/velero/pkg/apis/velero/v1#VolumeSnapshotLocationSpec so it would be consistent with the rest of the schema link references?

> |Location to store volume snapshots, as described in link:https://github.com/openshift/oadp-operator/blob/master/docs/config/bsl_and_vsl.md[Volume Snapshot Location Specification].

Reviewer: can we point this link to some downstream reference instead? I'm unsure if there is such doc downstream though

|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#SnapshotLocation[Complete schema definitions for the type `SnapshotLocation`].

.ApplicationConfig
[options="header"]
|===
|Property|Type|Description

|`velero`
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#VeleroConfig[VeleroConfig]
|Defines the configuration for the Velero server.

|`restic`
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#ResticConfig[ResticConfig]
|Defines the configuration for the Restic server.
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#ApplicationConfig[Complete schema definitions for the type `ApplicationConfig`].

.VeleroConfig
[options="header"]
|===
|Property|Type|Description

|`featureFlags`
|[] link:https://pkg.go.dev/builtin#string[string]
|Defines the list of features to enable for the Velero instance.

|`defaultPlugins`
|[] link:https://pkg.go.dev/builtin#string[string]

> |Five types of default Velero plug-ins can be installed: `AWS`,`Azure`, `CSI`, `GCP`, and `OpenShift`.

Reviewer: the plugin names should be lower cased, i.e, `openshift`, `aws`, `azure`, `csi`, `gcp`.
also missing additional default plugin option called `kubevirt`

|`customPlugins`
|[]link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#CustomPlugin[CustomPlugin]
|Used for installation of custom Velero plug-ins.

> link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/plugins.md[Plug-ins customization] gives more details on working with default and custom plug-ins.

Reviewer: could we point this link to the relevant downstream doc and not the one on the upstream?

|`restoreResourcesVersionPriority`
|link:https://pkg.go.dev/builtin#string[string]
|Represents a configmap that is created if defined for use in conjunction with the `EnableAPIGroupVersions` feature flag. Defining this field automatically adds `EnableAPIGroupVersions` to the Velero server feature flag.

|`noDefaultBackupLocation`
|link:https://pkg.go.dev/builtin#bool[bool]
|To install Velero without a default backup storage location, you must set the `noDefaultBackupLocation` flag in order to confirm installation.

|`podConfig`
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#PodConfig[`PodConfig`]
|Defines the configuration of the `Velero` pod.

|`logLevel`
|link:https://pkg.go.dev/builtin#string[string]

> |Velero server’s log level (use Debug for the most granular logging, leave unset for Velero default). Valid options are `trace`, `debug`, `info`, `warning`, `error`, `fatal`, and `panic`.

Reviewer: *use **d**ebug

|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#VeleroConfig[Complete schema definitions for the type `VeleroConfig`].

.CustomPlugin
[options="header"]
|===
|Property|Type|Description

|`name`
|link:https://pkg.go.dev/builtin#string[string]
|Name of custom plug-in.

|`image`
|link:https://pkg.go.dev/builtin#string[string]
|Image of custom plug-in.
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#CustomPlugin[Complete schema definitions for the type `CustomPlugin`].

.ResticConfig
[options="header"]
|===
|Property|Type|Description

|`enable`
|*link:https://pkg.go.dev/builtin#bool[bool]
|If set to `true`, enables backup and restore using Restic. If set to `false`, snapshots are needed.

|`supplementalGroups`
|[]link:https://pkg.go.dev/builtin#int64[int64]
|Defines the Linux groups to be applied to the `Restic` pod.

|`timeout`
|link:https://pkg.go.dev/builtin#string[string]
|Defines the Restic timeout. Default value is 1 hour.

> |`PodConfig`

Reviewer: should be camelCase, i.e, `podConfig`

.PodConfig
[options="header"]
|===
|Property|Type|Description

|`nodeSelector`
|map [ link:https://pkg.go.dev/builtin#string[string] ] link:https://pkg.go.dev/builtin#string[string]
|Defines the `nodeSelector` to be supplied to a `Velero` `podSpec` or a `Restic` `podSpec`.

|`tolerations`
|[]link:https://pkg.go.dev/k8s.io/api/core/v1#Toleration[ Corev1.Toleration]
|Defines the list of tolerations to be applied to a Velero deployment or a Restic `daemonset`.

|`resourceAllocations`
|link:https://pkg.go.dev/k8s.io/api/core/v1#ResourceRequirements[corev1.ResourceRequirements]
|Set specific resource `limits` and `requests` for a `Velero` pod or a `Restic` pod as described in link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/resource_req_limits.md[Resource Requests and Limits Customization].

If a `Velero` pod or a `Restic` pod fails or restarts due to a lack of memory or of CPU, link:https://docs.openshift.com/container-platform/4.10/backup_and_restore/application_backup_and_restore/troubleshooting.html#oadp-pod-crash-resource-request[you can set specific resource requests for either of those resources].

> |`Labels`

Reviewer: should be lowercase, i.e, `labels`

|map [ link:https://pkg.go.dev/builtin#string[string] ] link:https://pkg.go.dev/builtin#string[string]
|Labels to add to pods.
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#PodConfig[Complete schema definitions for the type `PodConfig`].

.Features
[options="header"]
|===
|Property|Type|Description

|`dataMover`

> |link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#DataMover[`DataMover`]

Reviewer: it's suppose to be a pointer, i.e: `*DataMover`

|Defines the configuration of the DPA Data Mover.
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#Features[Complete schema definitions for the type `Features`].

.DataMover
[options="header"]
|===
|Property|Type|Description

|`enable`

> |*link:https://pkg.go.dev/builtin#bool[bool]

Reviewer: not supposed to be a pointer (please remove the `*`)

|If set to `true`, deploys the volume snapshot mover controller and a modified CSI Data Mover plug-in. If set to `false`, these are not deployed.

|`credentialName`

> |*link:https://pkg.go.dev/builtin#string[string]

Reviewer: not supposed to be a pointer (please remove the `*`)

|`unsupportedOverrides`
|map [ link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#UnsupportedImageKey)[UnsupportedImageKey] ]  link:https://pkg.go.dev/builtin#string[string]
|Can be used to override the deployed dependent images for development. Options are `veleroImageFqin`, `awsPluginImageFqin`, `openshiftPluginImageFqin`, `azurePluginImageFqin`, `gcpPluginImageFqin`, `csiPluginImageFqin`, `dataMoverImageFqin`, `resticRestoreImageFqin`, `kubevirtPluginImageFqin`, and `operator-type`.

|`podAnnotations`
|map [ link:https://pkg.go.dev/builtin#string[string] ] link:https://pkg.go.dev/builtin#string[string]
|Used to add annotations to pods deployed by Operators.

|`podDnsPolicy`
|link:https://pkg.go.dev/k8s.io/api/core/v1#DNSPolicy[`corev1.DNSPolicy`]
|Defines the configuration of the DNS of a pod.

|`podDnsConfig`
|link:https://pkg.go.dev/k8s.io/api/core/v1#PodDNSConfig[`corev1.PodDNSConfig`]
|Defines the DNS parameters of a pod in addition to those generated from `DNSPolicy`.

|`backupImages`
|*link:https://pkg.go.dev/builtin#bool[bool]
|Used to specify whether or not you want to deploy a registry for enabling backup and restore of images.

|`configuration`

> |*link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#Applicationbucket   | [*CloudStorageLocation](https://pkg.go.dev/github.com/openshift/oadp-operator@master/api/v1alpha1#CloudStorageLocation) | [Tech Preview] Automate creation of bucket at some cloud storage providers for use as backup storage locationConfig[`ApplicationConfig`]

Reviewer 2: ```suggestion
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#Applicationbucket   | [*CloudStorageLocation](https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#CloudStorageLocation) | [Tech Preview] Automate creation of bucket at some cloud storage providers for use as backup storage locationConfig[`ApplicationConfig`]
```

|Used to define the data protection application's server configuration.

|`features`
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#Features[`Features`]
|Defines the configuration for the DPA to enable the Technology Preview features.
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#DataProtectionApplicationSpec[Complete schema definitions for the OADP API].

.BackupLocation
[options="header"]
|===
|Property|Type|Description

|`velero`
|*link:https://velero.io/docs/v1.9/api-types/backupstoragelocation/[velero.BackupStorageLocationSpec]
|Location to store volume snapshots, as described in link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/bsl_and_vsl.md[Backup Storage Locations and Volume Snapshot Locations Customization].

|`bucket`

> | [* link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#CloudStorageLocation[CloudStorageLocation] ]
> | [Technology Preview] Automate creation of bucket at some cloud storage providers for use as backup storage location.

Reviewer 2: ```suggestion
| [* link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#CloudStorageLocation[CloudStorageLocation] ]
| [Technology Preview] Automates creation of bucket at some cloud storage providers for use as backup storage location.
```

Reviewer 2: sorry about my typo earlier

Reviewer 2: ```suggestion
| [* link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#CloudStorageLocation[CloudStorageLocation] ]
| [Technology Preview] Automates creation of bucket at some cloud storage providers for use as a backup storage location.
```

> | [Technology Preview] Automate creation of bucket at some cloud storage providers for use as backup storage location.

Reviewer 2: ```suggestion
| [* link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#CloudStorageLocation[CloudStorageLocation] ]
| [Technology Preview] Automates creation of bucket at some cloud storage providers for use as backup storage location.
```

.DataProtectionApplicationSpec
[options="header"]
|===
|Property|Type|Description

|`backupLocations`
|[] link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#BackupLocation[`BackupLocation`]
|Defines the list of configurations to use for `BackupStorageLocations`.

|`snapshotLocations`
|[] link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#SnapshotLocation[`SnapshotLocation`]
|Defines the list of configurations to use for `VolumeSnapshotLocations`.

|`unsupportedOverrides`
|map [ link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#UnsupportedImageKey[UnsupportedImageKey] ]  link:https://pkg.go.dev/builtin#string[string]
|Can be used to override the deployed dependent images for development. Options are `veleroImageFqin`, `awsPluginImageFqin`, `openshiftPluginImageFqin`, `azurePluginImageFqin`, `gcpPluginImageFqin`, `csiPluginImageFqin`, `dataMoverImageFqin`, `resticRestoreImageFqin`, `kubevirtPluginImageFqin`, and `operator-type`.

|`podAnnotations`
|map [ link:https://pkg.go.dev/builtin#string[string] ] link:https://pkg.go.dev/builtin#string[string]
|Used to add annotations to pods deployed by Operators.

|`podDnsPolicy`

> |link:https://pkg.go.dev/k8s.io/api/core/v1#DNSPolicy[`corev1.DNSPolicy`]

Reviewer: maybe change the link's text to just `DNSPolicy` instead of corev1.DNSPolicy, as corev1 is not an actual package (it's an import alias). The package is actually v1, but I think it's redundant to mention it.

|Defines the configuration of the DNS of a pod.

|`podDnsConfig`

> |link:https://pkg.go.dev/k8s.io/api/core/v1#PodDNSConfig[`corev1.PodDNSConfig`]

Reviewer: maybe change the link's text to just `PodDNSConfig` instead of corev1.PodDNSConfig, as corev1 is not an actual package (it's an import alias). The package is actually v1, but I think it's redundant to mention it.

|`bucket`
| *link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#CloudStorageLocation[CloudStorageLocation]
| [Technology Preview] Automates creation of a bucket at some cloud storage providers for use as a backup storage location.

[IMPORTANT] +
==== +
The `bucket` parameter is a Technology Preview feature only. Technology Preview features are not supported with Red Hat production service level agreements (SLAs) and might not be functionally complete. Red Hat does not recommend using them in production. These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.

For more information about the support scope of Red Hat Technology Preview features, see link:https://access.redhat.com/support/offerings/techpreview/[Technology Preview].
====
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#BackupLocation[Complete schema definitions for the type `BackupLocation`].

.SnapshotLocation
[options="header"]
|===
|Property|Type|Description

|`velero`

> |*link:https://pkg.go.dev/github.com/vmware-tanzu/velero/pkg/apis/velero/v1#VolumeSnapshotLocationSpec/[VolumeSnapshotLocationSpec]

Reviewer: please just remove the backslash after v1#VolumeSnapshotLocationSpec, as it's not pointing to the correct hyperlink

|`supplementalGroups`
|[]link:https://pkg.go.dev/builtin#int64[int64]
|Defines the Linux groups to be applied to the `Restic` pod.

|`timeout`
|link:https://pkg.go.dev/builtin#string[string]
|A user-supplied duration string that defines the Restic timeout. Default value is `1hr` (1 hour). A duration string is a possibly signed sequence of decimal numbers, each with optional fraction and a unit suffix, such as `300ms`, -1.5h` or `2h45m`. Valid time units are `ns`, `us` (or `µs`), `ms`, `s`, `m`, and `h`.

|`podConfig`
|*link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#PodConfig[`PodConfig`]
|Defines the configuration of the `Restic` pod.
|===

link:https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#ResticConfig[Complete schema definitions for the type `ResticConfig`].

.PodConfig
[options="header"]
|===
|Property|Type|Description

|`nodeSelector`
|map [ link:https://pkg.go.dev/builtin#string[string] ] link:https://pkg.go.dev/builtin#string[string]
|Defines the `nodeSelector` to be supplied to a `Velero` `podSpec` or a `Restic` `podSpec`.

|`tolerations`

> |[]link:https://pkg.go.dev/k8s.io/api/core/v1#Toleration[ Corev1.Toleration]

Reviewer: maybe change the link's text to just `Toleration` instead of Corev1.Toleration, as Corev1 is not an actual package (it's an import alias).  The package is actually v1, but I think it's redundant to mention it.

|Defines the list of tolerations to be applied to a Velero deployment or a Restic `daemonset`.

|`resourceAllocations`

> |link:https://pkg.go.dev/k8s.io/api/core/v1#ResourceRequirements[corev1.ResourceRequirements]

Reviewer: maybe change the link's text to just  `ResourceRequirements` instead of corev1.ResourceRequirements as corev1 is not an actual package (it's an import alias). The package is actually v1, but I think it's redundant to mention it.

> |Set specific resource `limits` and `requests` for a `Velero` pod or a `Restic` pod as described in link:https://github.com/openshift/oadp-operator/blob/076f1495f6de65b0e96392cb6ef7cff66cf7defb/docs/config/resource_req_limits.md[Resource Requests and Limits Customization].

Reviewer: could you please refer some downstream doc instead of https://github.com/openshift/oadp-operator...?

Reviewer 2: This could work
https://access.redhat.com/documentation/en-us/openshift_container_platform/4.10/html-single/backup_and_restore/index#oadp-setting-resource-limits-and-requests_installing-oadp-aws

Reviewer 3: +1. We are not allowed to link to GitHub. See the [OCP doc guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#links-to-external-websites).

This needs to be an xref pointing to downstream docs.

> |Set specific resource `limits` and `requests` for a `Velero` pod or a `Restic` pod as described in link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.10/html-single/backup_and_restore/index#oadp-setting-resource-limits-and-requests_installing-oadp-aws[Setting Velero CPU and memory resource allocations].

Reviewer: is the OCP version of the doc should be hard-coded?

> If a `Velero` pod or a `Restic` pod fails or restarts due to a lack of memory or of CPU, link:https://docs.openshift.com/container-platform/4.10/backup_and_restore/application_backup_and_restore/troubleshooting.html#oadp-pod-crash-resource-request[you can set specific resource requests for either of those resources].

Reviewer: is the OCP version of the doc should be hard-coded?

Reviewer 3: No, it should not be hardcoded. This needs to be turned into an xref. However, xrefs in modules are not permitted. Any links (cross references) to OpenShift documentation needs to be in an assembly, probably in an "Additional Resources" section.
